README.md
<img align="right" width="200" src="https://github.com/olavoparno/olavoparno.github.io/raw/master/images/jest-badges-readme-logo4.png" />

# [Jest Badges Readme](https://www.npmjs.com/package/@olavoparno/jest-badges-readme)

[![CodeClimate](https://img.shields.io/codeclimate/maintainability/olavoparno/jest-badges-readme.svg)](https://codeclimate.com/github/olavoparno/jest-badges-readme) [![DevDeps](https://david-dm.org/olavoparno/jest-badges-readme/dev-status.svg)](https://david-dm.org/olavoparno/jest-badges-readme/dev-status.svg) [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/olavoparno/jest-badges-readme/blob/master/LICENSE) [![npm version](https://img.shields.io/npm/v/@olavoparno/jest-badges-readme.svg?style=flat)](https://www.npmjs.com/package/@olavoparno/jest-badges-readme) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/olavoparno/jest-badges-readme/pulls) ![BuiltBy](https://img.shields.io/badge/TypeScript-Lovers-black.svg "img.shields.io") ![ForTheBadge](https://img.shields.io/badge/Using-Badges-red.svg "ForTheBadge")

### Creates Jest Coverage badges for your README.md file

| Statements | Branches | Functions | Lines | Build Status |
| -----------|----------|-----------|-------| ------------ |
| ![Statements](https://img.shields.io/badge/Coverage-96.89%25-brightgreen.svg "Make me better!") | ![Branches](https://img.shields.io/badge/Coverage-94.59%25-brightgreen.svg "Make me better!") | ![Functions](https://img.shields.io/badge/Coverage-100%25-brightgreen.svg "Make me better!") | ![Lines](https://img.shields.io/badge/Coverage-96.89%25-brightgreen.svg "Make me better!") | ![BuildStatus](https://img.shields.io/badge/Build-Passing-brightgreen.svg "Building Status") |

### TL;DR

- To be short, all it does is: **Creates your README.md dynamically with your Jest Coverage Badges based on your README-template.md file**

### Installation

- Global Installation
  ```bash
    npm i -g @olavoparno/jest-badges-readme
  ```

- Local Installation
  ```bash
    npm i @olavoparno/jest-badges-readme --save-dev
  ```

### Configuration and requirements

  1. Have Jest installed and with its Coverage Reporting configuration as below
  
  - It may be set inside your Jest config within `package.json` or inside your jest config file i.e. `jest.config.js` or `jestconfig.json`

    ```js
      "coverageReporters": [
        "json-summary",
        "lcov"
      ]
    ```
  
  2. Have a _*README-template.md*_ file created with the following template (please do feel free to change its presentation structure, e.g. in a table or not)

  ```md
  | Statements | Branches | Functions | Lines |
  | -----------|----------|-----------|-------|
  | ![Statements](#statements# "Make me better!") | ![Branches](#branches# "Make me better!") | ![Functions](#functions# "Make me better!") | ![Lines](#lines# "Make me better!") |
  ```

  3. Call if from your terminal or from your NPM/YARN Scripts
 
  - For locally installed library

    ```json
      "make-badges": "node_modules/.bin/jest-badges-readme"
    ```

  - For globally installed library

    ```json
      "make-badges": "npm run jest-badges-readme"
    ```

  - You may also provide a different coverage folder path passed as arguments

    ```json
      "make-badges": "npm run jest-badges-readme --coverageDir='./my-custom-coverage-folder'"
    ```

  - And finally via the very terminal

    ```bash
      jest-badges-readme --coverageDir='./my-custom-coverage-folder'
    ```

### Short summary onto what Jest and Jest Badges Readme covers for us

  - % Stmts (which states for Statements)
  - % Branch (pretty self explanatory)
  - % Functions (won't repeat myself)
  - % Lines (last but not least)

#### This library will create badges URLs for these former 4 items into your _*README-template.md*_ if you followed the previous steps

  - Open up your _*README-template.md*_ and add 4 _*markers*_ inside it for each Jest coverage item respectively. These _*markers*_ must be as follows:

    1. '#statements#'
    2. '#branches#'
    3. '#functions#'
    4. '#lines#'

#### If you feel you might need a local building badge indicator, add these NPM scripts in order to manage your local building badge as well

  ```json
    "prebuild": "echo nok > .buildstatus",
    "postbuild": "echo ok > .buildstatus"
  ```

  - Add one more marker like the example below. This is going to create a file called _*.buildstatus*_ in your project's root telling this library if your build is either successful or a failure
    1. '#buildstatus#'
  
#### Example of _*README-template.md*_ with building status too

  ```md
  | Statements | Branches | Functions | Lines | Build Status |
  | -----------|----------|-----------|-------| ------------ |
  | ![Statements](#statements# "Make me better!") | ![Branches](#branches# "Make me better!") | ![Functions](#functions# "Make me better!") | ![Lines](#lines# "Make me better!") | ![BuildStatus](#buildstatus# "Building Status") |
  ```

### Using as a part of your githooks
  
  - If you want to have this run on the precommit hook and update the commit in place, just install husky and add the `precommit` script to your package.json

  1. Install Husky
   
  ```bash
    npm install --save-dev husky 
  ```

  2. Add your precommit script
    
  ```json
    "husky": {
      "hooks": {
        "pre-commit": "jest && node_modules/.bin/jest-badges-readme && git add 'README.md'"
      }
    }
  ```

  3. Git Commit and Push. Just use your workflow as usual. If your tests fail, no commit. If they pass, update the README.md and add the file to the commit. Nice!

### Contributing

The purpose of this library is to motivate developers to constantly write a solid documentation and testing. A complete and nice looking documentation is key to a successful development and code maintainability.
Please feel free to open any issues you might come up with and to submit your own PRs. There is not a contributing guide yet

### Contributors

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://olavoparno.github.io"><img src="https://avatars1.githubusercontent.com/u/7513162?v=4" width="70px;" alt=""/><br /><sub><b>Olavo Parno</b></sub></a><br /><a href="#ideas-olavoparno" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/olavoparno/jest-badges-readme/commits?author=olavoparno" title="Code">💻</a> <a href="https://github.com/olavoparno/jest-badges-readme/commits?author=olavoparno" title="Tests">⚠️</a></td>
    <td align="center"><a href="https://github.com/nothingismagick"><img src="https://avatars1.githubusercontent.com/u/35242872?v=4" width="70px;" alt=""/><br /><sub><b>nothingismagick</b></sub></a><br /><a href="#ideas-nothingismagick" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/olavoparno/jest-badges-readme/issues?q=author%3Anothingismagick" title="Bug reports">🐛</a> <a href="#content-nothingismagick" title="Content">🖋</a></td>
    <td align="center"><a href="http://www.fallenclient.co.uk"><img src="https://avatars2.githubusercontent.com/u/326470?v=4" width="70px;" alt=""/><br /><sub><b>Dave Fisher</b></sub></a><br /><a href="https://github.com/olavoparno/jest-badges-readme/issues?q=author%3Afallenclient" title="Bug reports">🐛</a></td>
    <td align="center"><a href="http://twitter.com/zaggino"><img src="https://avatars1.githubusercontent.com/u/1067319?v=4" width="70px;" alt=""/><br /><sub><b>Martin Zagora</b></sub></a><br /><a href="#ideas-zaggino" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/olavoparno/jest-badges-readme/issues?q=author%3Azaggino" title="Bug reports">🐛</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

### License

Jest Badges Readme is [MIT licensed](./LICENSE).

<div style="display: none; visibility: hidden; opacity: 0;">
  <img height=0 width=0 style="display: none; visibility: hidden; opacity: 0;" src="http://hits.dwyl.com/olavoparno/jest-badges-readme.svg" />
</div>


  


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.5.1](https://github.com/olavoparno/jest-badges-readme/compare/v1.5.0...v1.5.1) (2020-07-08)


### Bug Fixes

* **project:** remove logging color and updates ([9470344](https://github.com/olavoparno/jest-badges-readme/commit/947034406aeda2215b88a7f54ea3e89f46b05bc6))
* **project:** remove logging color and updates ([831db61](https://github.com/olavoparno/jest-badges-readme/commit/831db610a54a9a0825da324a4db86a96aa4c1c04))

# [1.5.0](https://github.com/olavoparno/jest-badges-readme/compare/v1.4.2...v1.5.0) (2020-07-07)



## [1.4.2](https://github.com/olavoparno/jest-badges-readme/compare/v1.4.1...v1.4.2) (2020-06-20)



## [1.4.1](https://github.com/olavoparno/jest-badges-readme/compare/v1.4.0...v1.4.1) (2020-02-13)



# [1.4.0](https://github.com/olavoparno/jest-badges-readme/compare/v1.3.8...v1.4.0) (2019-12-20)



## [1.3.8](https://github.com/olavoparno/jest-badges-readme/compare/v1.3.7...v1.3.8) (2019-12-11)



## [1.3.7](https://github.com/olavoparno/jest-badges-readme/compare/v1.3.6...v1.3.7) (2019-09-01)



## [1.3.6](https://github.com/olavoparno/jest-badges-readme/compare/v1.3.5...v1.3.6) (2019-06-08)



## [1.3.5](https://github.com/olavoparno/jest-badges-readme/compare/1.3.4...v1.3.5) (2019-06-06)



## [1.3.4](https://github.com/olavoparno/jest-badges-readme/compare/1.3.3...1.3.4) (2019-04-05)



## [1.3.3](https://github.com/olavoparno/jest-badges-readme/compare/1.3.2...1.3.3) (2019-04-03)



## [1.3.2](https://github.com/olavoparno/jest-badges-readme/compare/1.3.1...1.3.2) (2019-03-29)



## [1.3.1](https://github.com/olavoparno/jest-badges-readme/compare/1.3.0...1.3.1) (2019-03-29)



# [1.3.0](https://github.com/olavoparno/jest-badges-readme/compare/1.2.9...1.3.0) (2019-03-25)



## [1.2.9](https://github.com/olavoparno/jest-badges-readme/compare/1.2.8...1.2.9) (2019-03-25)



## [1.2.8](https://github.com/olavoparno/jest-badges-readme/compare/1.2.7...1.2.8) (2019-03-21)



## [1.2.7](https://github.com/olavoparno/jest-badges-readme/compare/1.2.6...1.2.7) (2019-03-15)



## [1.2.6](https://github.com/olavoparno/jest-badges-readme/compare/v1.1.5...1.2.6) (2019-02-02)


### Bug Fixes

* **buildBadge:** adjust badge color ([993737e](https://github.com/olavoparno/jest-badges-readme/commit/993737e))
* **buildstatus:** change test string ([519e0d3](https://github.com/olavoparno/jest-badges-readme/commit/519e0d3))


### Features

* **badges:** add building badges feature ([285937b](https://github.com/olavoparno/jest-badges-readme/commit/285937b))



## [1.1.5](https://github.com/olavoparno/jest-badges-readme/compare/1490183...v1.1.5) (2018-11-24)


### Bug Fixes

* **package:** fix cli.js script ([1490183](https://github.com/olavoparno/jest-badges-readme/commit/1490183))


### Features

* **project:** add error handling ([f533869](https://github.com/olavoparno/jest-badges-readme/commit/f533869))


src/tests/creator.spec.ts
import * as fs from 'fs';

import Creator from '../helper';

let creator: Creator;
let readmeFile: string;
let readmeTemplateFile: string;
let fakeReadMeFile: string;
let fakeCoverageNonExistent: string;
let fakeCoverageZeroLength: string;
let fakeCoverageBadInfo: string;
let fakeCoverageLowValue: string;
let trueBuildStatus: string;
let falseBuildStatus: string;
let fakeBuildStatus: string;
const reportKeys = [
  { key: 'branches', value: '![Branches]' },
  { key: 'functions', value: '![Functions]' },
  { key: 'lines', value: '![Lines]' },
  { key: 'statements', value: '![Statements]' },
];

describe('Badge Maker', () => {
  beforeEach(() => {
    creator = new Creator();
  });
  beforeAll(() => {
    readmeFile = './README.md';
    readmeTemplateFile = './README-template.md';
    fakeReadMeFile = './FAKEREADME.md';
    fakeCoverageNonExistent = './FAKEJSON.json';
    fakeCoverageZeroLength = 'src/tests/mock/fakeJsonZeroLength.json';
    fakeCoverageBadInfo = 'src/tests/mock/fakeJsonBadInfo.json';
    fakeCoverageLowValue = 'src/tests/mock/fakeJsonLowValue.json';
    trueBuildStatus = 'src/tests/mock/.trueBuildStatus';
    falseBuildStatus = 'src/tests/mock/.falseBuildStatus';
    fakeBuildStatus = './.FAKEBUILD';
  });
  test('checks README.md from template file creation', () => {
    expect(creator.createReadme(readmeTemplateFile)).toBeTruthy();
    expect(fs.existsSync(readmeFile)).toBeTruthy();
  });
  test('checks README.md from readme file creation', () => {
    expect(creator.createReadme()).toBeTruthy();
    expect(fs.existsSync(readmeFile)).toBeTruthy();
  });
  test('calls insertBadges', () => {
    if (creator.createReadme()) {
      expect(creator.insertBadges()).toBeTruthy();
    }
  });
  test('calls failing insertBadges non existent', () => {
    if (creator.createReadme()) {
      expect(creator.insertBadges(fakeCoverageNonExistent)).toBeFalsy();
    }
  });
  test('calls failing insertBadges zero length', () => {
    if (creator.createReadme()) {
      expect(creator.insertBadges(fakeCoverageZeroLength)).toBeFalsy();
    }
  });
  test('calls failing insertBadges bad info', () => {
    if (creator.createReadme()) {
      expect(creator.insertBadges(fakeCoverageBadInfo)).toBeFalsy();
    }
  });
  test('calls insertBadges with low values', () => {
    if (creator.createReadme()) {
      expect(creator.insertBadges(fakeCoverageLowValue)).toBeTruthy();
    }
  });
  test('checks if all lines were updated', () => {
    fs.readFile(readmeFile, (err, data) => {
      if (err) {
        throw err;
      }
      reportKeys.forEach((reportKey: Record<string, string>) => {
        expect(data.indexOf(reportKey.key)).toBeGreaterThan(0);
      });
    });
  });
  describe('Testing building', () => {
    test('check build status failing', () => {
      expect(creator.getBuildStatus(falseBuildStatus)).toBeFalsy();
    });
    test('check build status nonExistent', () => {
      expect(creator.getBuildStatus(fakeBuildStatus)).toBeFalsy();
    });
    test('check build status passing', () => {
      expect(creator.getBuildStatus(trueBuildStatus)).toBeTruthy();
    });
    afterAll(() => {
      creator.createReadme();
    });
  });
  describe('Testing with arguments', () => {
    beforeAll(() => {
      process.argv.push('--coverageDir="./coverage"');
    });
    test('checks for different coverageDirectory as args', () => {
      if (creator.createReadme()) {
        expect(creator.insertBadges()).toBeFalsy();
      }
    });
  });
});


cli.js
#!/usr/bin/env node

const fs = require('fs')
const replace = require('replace-in-file')

try {
  fs.copyFileSync('./README-template.md', 'README.md');
  console.log("\x1b[32m", 'Template read succesfully. README.md created!'); 
} catch (error) {
  console.log("\x1b[30m", error);
  console.log("\x1b[32m", 'You must have a README-template.md created. Please read the documentation.');
}

const getColour = coverage => {
  if (coverage < 80) {
    return 'red';
  }
  if (coverage < 90) {
    return 'yellow';
  }
  return 'brightgreen';
}

const reportKeys = ['lines', 'statements', 'functions', 'branches'];

const getBadge = (report, key) => {
  if (!(report && report.total && report.total[key])) {
    throw new Error('Malformed coverage report.');
  }
  const coverage = report.total[key].pct;
  const colour = getColour(coverage);
  return `https://img.shields.io/badge/Coverage-${coverage}${encodeURI('%')}-${colour}.svg`;
}

try {
  fs.readFile('./coverage/coverage-summary.json', 'utf8', (err, res) => {
    if (err) throw err
    const report = JSON.parse(res);
    let url;
    let options;
    let pattern;
    reportKeys.forEach(key => {
      url = getBadge(report, key);
      pattern = "#" + key + "#";
      console.log(url)
      options = {
        files: './README.md',
        from: pattern,
        to: url,  
      };
      try {
        const changes = replace.sync(options);
        console.log("\x1b[32m", 'Template read succesfully. README.md created!'); 
        console.log('Modified pattern: ', pattern, ' with: ', changes.join(', '));
      }
      catch (error) {
        console.log("\x1b[30m", error);
        console.error("\x1b[32m", 'An error occurred when filling your README.md. Care to check it?');
      }
    });
  });
} catch (error) {
  console.log("\x1b[30m", error);
  console.log("\x1b[32m", 'Do you have Jest running? If so, is it properly configured?');
}

src/helper/index.ts
import * as fs from 'fs';

import { badgeHashs } from './constants';
import { IReport } from './interface';

const readmeFile: string = badgeHashs.readmeFile;
const readmeTemplateFile: string = badgeHashs.readmeTemplateFile;

export default class Helper {
  private reportKeys: Record<string, string>[] = badgeHashs.hashes.jest;

  public createReadme = (path: string = readmeTemplateFile): boolean => {
    if (fs.existsSync(path)) {
      fs.copyFileSync(path, readmeFile);
      console.log(`Template read succesfully. ${readmeFile} created!`);
      return true;
    }
    if (fs.existsSync(readmeFile)) {
      console.log(`Readme read succesfully. ${readmeFile} will be used accordingly!`);
      return true;
    }
    console.log(`Error: ${readmeTemplateFile} OR ${readmeFile} files were not found.`);
    console.log(
      `You must have a ${readmeTemplateFile} OR a valid ${readmeFile} file created. Please read the documentation.`,
    );
    return false;
  };

  public insertBadges = (path: string = badgeHashs.jsonCoverageFile): boolean => {
    let returnValue: boolean = false;
    const coveragePath = this.getCoveragePath(path);
    if (fs.existsSync(coveragePath)) {
      const coverageFile = fs.readFileSync(coveragePath, 'utf8');
      if (!coverageFile.length) {
        console.log('Malformed coverage report. Please run Jest again');
        return false;
      }
      console.log(`Reading coverage file from ${coveragePath}`);
      const report: IReport = JSON.parse(coverageFile);
      let readmeFileData: string = fs.readFileSync(readmeFile, 'utf8');
      let hadChange = 0;
      this.reportKeys.forEach((reportKey) => {
        const url: string = this.getBadge(report, reportKey.key);
        if (url.length) {
          const pattern: string = reportKey.value;

          const startIndex = readmeFileData.indexOf(pattern);
          const valueToChangeStart = readmeFileData.slice(startIndex + pattern.length + 1);

          const valueToChangeIndex = valueToChangeStart.indexOf(')');
          const valueToChangeFinal = valueToChangeStart.substring(0, valueToChangeIndex);

          const newUrl = `${url} "Make me better!"`;
          const isItTheSame = valueToChangeFinal.localeCompare(newUrl);

          readmeFileData = readmeFileData.replace(valueToChangeFinal, newUrl);

          if (isItTheSame !== 0) {
            console.log(`Badge for ${reportKey.key} updated with: ${url}`);
            hadChange = 1;
          }

          returnValue = true;
          fs.writeFileSync(readmeFile, readmeFileData, 'utf8');
        } else {
          returnValue = false;
        }
      });
      if (hadChange === 0) {
        console.log('No update on coverage!');
      }
      this.getBuildStatus();
    } else {
      console.log(`Error: ${coveragePath} file not found. Is it at the default location?`);
      console.log('Do you have Jest installed? If so, is it properly configured? If you do, then run me');
      console.log("by passing args i.e. npm run jest-badges-readme --coverageDir='my-custom-coverage-folder'");
      returnValue = false;
    }
    return returnValue;
  };

  public getBuildStatus = (path: string = badgeHashs.buildFile): boolean => {
    let url: string = badgeHashs.getBuildUrl('Build', 'Passing', 'brightgreen');
    let returnValue: boolean = false;
    const pattern: string = badgeHashs.hashes.build.value;
    if (fs.existsSync(path)) {
      let readmeFileData = fs.readFileSync(readmeFile, 'utf8');
      const buildStatus = fs.readFileSync(path, 'utf8');
      if (buildStatus === 'ok') {
        url = badgeHashs.getBuildUrl('Build', 'Passing', 'brightgreen');
        returnValue = true;
      } else if (buildStatus === 'nok') {
        url = badgeHashs.getBuildUrl('Build', 'Failing', 'brightred');
        returnValue = false;
      }

      const startIndex = readmeFileData.indexOf(pattern);
      const valueToChangeStart = readmeFileData.slice(startIndex + pattern.length + 1);

      const valueToChangeIndex = valueToChangeStart.indexOf(')');
      const valueToChangeFinal = valueToChangeStart.substring(0, valueToChangeIndex);

      const newUrl = `${url} "Building Status"`;

      readmeFileData = readmeFileData.replace(valueToChangeFinal, newUrl);

      fs.writeFileSync(readmeFile, readmeFileData, 'utf8');
      console.log(`Writing building status ${buildStatus}`);
    } else {
      returnValue = false;
    }
    return returnValue;
  };

  private getCoveragePath = (path: string): string => {
    let coveragePath: string = path;
    let argPath: string = '';
    const args = process.argv
      .filter((item) => {
        if (item.indexOf('coverage') >= 0) {
          return item;
        }
      })
      .toString();

    if (args) {
      argPath = args.replace('--coverageDir=', '');
    }

    if (argPath && args.length > 0) {
      coveragePath = `${argPath}/coverage-summary.json`;
    }

    return coveragePath;
  };

  private getBadge = (report: IReport, key: string): string => {
    if (!(report && report.total && report.total[key])) {
      console.log(`Malformed coverage report for the key ${key}. Please run Jest again.`);
      return '';
    }
    const coverage: number = report.total[key].pct;
    const color = this.getColor(coverage);
    return badgeHashs.getCoverageUrl(coverage, color);
  };

  private getColor = (coverage: number): string => {
    if (coverage < 80) {
      return 'red';
    }
    if (coverage < 90) {
      return 'yellow';
    }
    return 'brightgreen';
  };
}
